src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
 
         private void DeterminePropertyName(JsonSerializerOptions options)
         {
-            if (PropertyInfo != null)
+            if (PropertyInfo == null)
             {
-                JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>();
-                if (nameAttribute != null)
-                {
-                    NameAsString = nameAttribute.Name;
+                return;
+            }
 
-                    // null is not valid; JsonClassInfo throws an InvalidOperationException after this return.
-                    if (NameAsString == null)
-                    {
-                        return;
-                    }
-                }
-                else if (options.PropertyNamingPolicy != null)
-                {
-                    NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
+            JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>(PropertyInfo);
+            if (nameAttribute != null)
+            {
+                NameAsString = nameAttribute.Name;
 
-                    // null is not valid; JsonClassInfo throws an InvalidOperationException after this return.
-                    if (NameAsString == null)
-                    {
-                        return;
-                    }
-                }
-                else
+                if (NameAsString == null)
                 {
-                    NameAsString = PropertyInfo.Name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else if (options.PropertyNamingPolicy != null)
+            {
+                NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
 
-                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
-                _name = Encoding.UTF8.GetBytes(NameAsString);
-
-                // Set the compare name.
-                if (options.PropertyNameCaseInsensitive)
+                if (NameAsString == null)
                 {
-                    NameUsedToCompareAsString = NameAsString.ToUpperInvariant();
-                    _nameUsedToCompare = Encoding.UTF8.GetBytes(NameUsedToCompareAsString);
-                }
-                else
-                {
-                    NameUsedToCompareAsString = NameAsString;
-                    _nameUsedToCompare = _name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else
+            {
+                NameAsString = PropertyInfo.Name;
```

Comment:
And `PropertyInfo.Name` is guaranteed to not be null, correct?

Reply:
Yes. The beginning of the method checks for it.